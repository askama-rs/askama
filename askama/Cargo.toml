[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
