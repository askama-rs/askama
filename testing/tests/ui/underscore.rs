use askama::Template;

#[derive(Template)]
#[template(source = r#"{% let x = [_] %}"#, ext = "html")]
struct UnderscoreErr1;

#[derive(Template)]
#[template(source = r#"{% if (_ + 12) != 0 %}{% endif %}"#, ext = "html")]
struct UnderscoreErr2;

#[derive(Template)]
#[template(source = r#"{% if 12 == _ %}{% endif %}"#, ext = "html")]
struct UnderscoreErr3;

#[derive(Template)]
#[template(source = r#"{% match _ %}{% endmatch %}"#, ext = "html")]
struct UnderscoreErr4;

#[derive(Template)]
#[template(source = r#"{% let _ %}"#, ext = "html")]
struct UnderscoreErr5;

#[derive(Template)]
#[template(source = r#"{% create _ %}"#, ext = "html")]
struct UnderscoreErr6;

fn main() {}
