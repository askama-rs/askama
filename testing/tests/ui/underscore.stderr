error: reserved keyword `_` cannot be used here
 --> <source attribute>:1:12
       "_] %}"
 --> tests/ui/underscore.rs:4:21
  |
4 | #[template(source = r#"{% let x = [_] %}"#, ext = "html")]
  |                     ^^^^^^^^^^^^^^^^^^^^^^

error: reserved keyword `_` cannot be used here
 --> <source attribute>:1:7
       "_ + 12) != 0 %}{% endif %}"
 --> tests/ui/underscore.rs:8:21
  |
8 | #[template(source = r#"{% if (_ + 12) != 0 %}{% endif %}"#, ext = "html")]
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: reserved keyword `_` cannot be used here
 --> <source attribute>:1:12
       "_ %}{% endif %}"
  --> tests/ui/underscore.rs:12:21
   |
12 | #[template(source = r#"{% if 12 == _ %}{% endif %}"#, ext = "html")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: reserved keyword `_` cannot be used here
 --> <source attribute>:1:9
       "_ %}{% endmatch %}"
  --> tests/ui/underscore.rs:16:21
   |
16 | #[template(source = r#"{% match _ %}{% endmatch %}"#, ext = "html")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: when you forward-define a variable, you cannot use a pattern in place of a variable name
 --> <source attribute>:1:7
       "_ %}"
  --> tests/ui/underscore.rs:20:21
   |
20 | #[template(source = r#"{% let _ %}"#, ext = "html")]
   |                     ^^^^^^^^^^^^^^^^

error: `_` cannot be used when there is no value assigned, use `let` instead
 --> UnderscoreErr6.html:1:10
       "_ %}"
  --> tests/ui/underscore.rs:24:21
   |
24 | #[template(source = r#"{% create _ %}"#, ext = "html")]
   |                     ^^^^^^^^^^^^^^^^^^^
